### ISSUE TITLE
Access: Rev...invalidates channels of ...

### ISSUE THREAD

**Issue by npomfret** [couchbase/sync_gateway#1791]
I wasn't sure where to post [this](https://github.com/couchbase/couchbase-lite-ios/issues/1256) question so I put it on the IOS project, but was wondering if someone here could help explain.

I'm seeing some channel revoking happen and couldn't figure out from the docs why it was happening.  I'm deleting a document and in my SG logs I'm seeing this message:

```
2016-05-23T11:00:50.240+01:00 Access: Rev "-3J2MSLICp3mnENpkblYOy3"/"2-1c3ac9136429a53ce9df8ec8df228c8a" invalidates channels of [c4ed2a23-a70e-4acb-8912-b22aa6751bf1]

2016-05-23T11:00:50.242+01:00 Access: Invalidate access of "c4ed2a23-a70e-4acb-8912-b22aa6751bf1"
```

And as a result (I think) my IOS client is not getting notified of a document deletion.  I don't think it's a bug, I just don't understand why its happening and how I can work around it.  Obviously I want the client to be updated when documents are deleted.

**Comment by npomfret** [comment]
Am I correct in thinking that there is no way to delete certain documents so that clients are notified of the delete?

The situation I have is that 1 document is created and during creation the sync_function assigns the document to a channel and some users to the same channel.  When that document is deleted, the users are removed from the channel BEFORE the information about the delete is sent to them, so they are never notified in their `_changes` feed about the deleted document?

**Comment by adamcfraser** [comment]
As @snej mentioned on the iOS issue, the forums are probably the right place for this kind of question.  

Currently Sync Gateway has the following behaviour:
1. When a document is removed from a channel, users who belong to that channel are notified that the document has been removed.
2. When a user is removed from a channel, that user does **not** receive removal notifications for all documents in that channel.  This is a long-standing open issue (https://github.com/couchbase/sync_gateway/issues/264) - the main obstacle being that it's difficult and expensive to identify the set of documents that a given user has access to only through that channel (and not via some other channel), and to handle this scenario for filtered replications.  

In your situation, there are a couple of options:
1. Define your channels per user, instead of per document, and assign the document to the users' channels, instead of the document's channels.
2. When you delete the document, maintain the access grants (usually done by referencing oldDoc in your sync function).

**Comment by npomfret** [comment]
Thanks for the reply.

I tried option 2, to use the oldDoc instead of the newDoc inside the sync-function in the case of deletes.  It didn't work.  The user who originally authored the doc never gets notified of the delete.  I guess they lose access to the channel before the update gets sent to them?

**Comment by adamcfraser** [comment]
@npomfret My apologies - I'd forgotten that access grants are ignored on tombstones.  I was thinking of the handling for channel assignment when I suggested option 2.  If you want to do something like option 2, you'd actually have to remove the access grants in a non-deleted revision, then subsequently tombstone the document.  

I'd like to better support this channel modeling pattern - I've filed #1796 to try to come up with an enhancement on the Sync Gateway side.
